File: platforms/hermes-agent/plugin/enigmagent/tools.py
"""
EnigmAgent Hermes Plugin — agent-callable tools.

Two tools are exposed to the Hermes agent:

    enigmagent_vault_status  — check if the vault server is running & unlocked
    enigmagent_vault_list    — list secret names and domains (NEVER values)

Both follow the Hermes tool handler convention:
    - Accept only simple Python types (str, int, etc.)
    - Return a JSON-encoded string
    - Include "error" key on failure

These tools intentionally NEVER return actual secret values.
Secrets are injected only through the resolver's pre_tool_call hook.
"""

from __future__ import annotations

import json
import logging
from typing import Callable

from .vault_client import VaultClient, VaultError

logger = logging.getLogger(__name__)


# ── Schema constants ──────────────────────────────────────────────────────────

VAULT_STATUS_SCHEMA: dict = {
    "type": "object",
    "properties": {},
    "required": [],
    "additionalProperties": False,
}

VAULT_LIST_SCHEMA: dict = {
    "type": "object",
    "properties": {},
    "required": [],
    "additionalProperties": False,
}


# ── Tool handler factories ────────────────────────────────────────────────────


def make_vault_status_handler(client: VaultClient) -> Callable:
    """
    Return a Hermes-compatible handler for ``enigmagent_vault_status``.

    The handler checks whether the EnigmAgent server is reachable and
    whether the vault is currently unlocked.  It never returns secret values.
    """

    def vault_status_handler(task_id: str | None = None, **kwargs) -> str:
        """Check EnigmAgent vault server status."""
        try:
            status = client.get_status()
            unlocked = bool(status.get("unlocked", False))

            if unlocked:
                message = (
                    "Vault is running and UNLOCKED. "
                    "All {{PLACEHOLDER}} references in tool arguments will be "
                    "resolved automatically before execution."
                )
            else:
                message = (
                    "Vault server is running but LOCKED. "
                    "Restart the server to unlock it:\n"
                    "  enigmagent-mcp --mode rest --port 3737 "
                    "--vault ~/.enigmagent/vault.json"
                )

            return json.dumps(
                {
                    "running": True,
                    "unlocked": unlocked,
                    "message": message,
                },
                ensure_ascii=False,
            )

        except VaultError as exc:
            if exc.code == "server_unreachable":
                return json.dumps(
                    {
                        "running": False,
                        "unlocked": False,
                        "message": (
                            "EnigmAgent server is NOT running. "
                            "Start it with:\n"
                            "  enigmagent-mcp --mode rest --port 3737 "
                            "--vault ~/.enigmagent/vault.json\n\n"
                            "For setup instructions run:\n"
                            "  hermes enigmagent-setup"
                        ),
                    },
                    ensure_ascii=False,
                )
            return json.dumps({"error": exc.code, "message": str(exc)})

        except Exception as exc:  # noqa: BLE001
            logger.exception("Unexpected error in vault_status")
            return json.dumps({"error": "unexpected", "message": str(exc)})

    return vault_status_handler


def make_vault_list_handler(client: VaultClient) -> Callable:
    """
    Return a Hermes-compatible handler for ``enigmagent_vault_list``.

    Lists all secret names and domain bindings stored in the vault.
    NEVER returns secret values — only metadata.

    The agent should use this to discover available placeholder names
    before constructing tool calls that need credentials.
    """

    def vault_list_handler(task_id: str | None = None, **kwargs) -> str:
        """List available EnigmAgent secrets (names and domains only — no values)."""
        try:
            entries = client.list_secrets()

            if not entries:
                return json.dumps(
                    {
                        "count": 0,
                        "entries": [],
                        "hint": (
                            "No secrets in vault. Add one with:\n"
                            "  enigmagent add SECRET_NAME @localhost <value>\n"
                            "  enigmagent add LOGIN:github.com @localhost <password>\n"
                            "  enigmagent add DOC_report.md @localhost <document-text>"
                        ),
                    },
                    ensure_ascii=False,
                )

            # Strip values — return only metadata.
            safe_entries = [
                {
                    "name":    e.get("name"),
                    "domain":  e.get("domain"),
                    "created": e.get("created"),
                }
                for e in entries
            ]

            return json.dumps(
                {
                    "count":   len(safe_entries),
                    "entries": safe_entries,
                    "usage": (
                        "Reference secrets in tool arguments as {{NAME}}. "
                        "They will be resolved automatically before the tool runs. "
                        "Examples:\n"
                        "  {{GITHUB_TOKEN}}  →  in bash: "
                        'curl -H "Authorization: Bearer {{GITHUB_TOKEN}}" ...\n'
                        "  {{LOGIN:github.com}}  →  in forms: password field\n"
                        "  {{DOC:report.md}}  →  inlines full document text"
                    ),
                },
                ensure_ascii=False,
            )

        except VaultError as exc:
            if exc.code == "vault_locked":
                return json.dumps(
                    {
                        "error": "vault_locked",
                        "message": (
                            "Vault is locked. Restart the server to unlock:\n"
                            "  enigmagent-mcp --mode rest --port 3737 "
                            "--vault ~/.enigmagent/vault.json"
                        ),
                    }
                )
            return json.dumps({"error": exc.code, "message": str(exc)})

        except Exception as exc:  # noqa: BLE001
            logger.exception("Unexpected error in vault_list")
            return json.dumps({"error": "unexpected", "message": str(exc)})

    return vault_list_handler

File: platforms/openclaw/extensions/enigmagent/src/tools.ts
/**
 * EnigmAgent OpenClaw Plugin — agent-callable tools.
 *
 * These tools let an agent inspect the vault's state and list available
 * secrets. They intentionally NEVER return actual secret values — secrets
 * are only injected inline via the middleware (src/middleware.ts).
 *
 * Tools exposed:
 *   enigmagent_vault_status  — check if vault server is running & unlocked
 *   enigmagent_vault_list    — list secret names and domains (no values)
 *
 * Usage in agent prompts:
 *   "Check if the vault is ready before attempting to push code."
 *   "List the available secrets so I know what credentials I have access to."
 *
 * Compatible with the OpenClaw AgentTool interface from
 * @openclaw/plugin-sdk/tools.
 */

import type { VaultStatusToolResult, VaultListToolResult } from './types.js';
import { VaultError } from './types.js';
import type { VaultClient } from './vault-client.js';

// ── Tool definitions ──────────────────────────────────────────────────────────

/**
 * OpenClaw AgentTool shape.
 * Matches @openclaw/plugin-sdk/tools AgentTool<TInput, TOutput>.
 */
export interface AgentTool<TInput = Record<string, unknown>, TOutput = unknown> {
  name: string;
  description: string;
  inputSchema: {
    type: 'object';
    properties: Record<string, unknown>;
    required?: string[];
  };
  execute(input: TInput): Promise<TOutput>;
}

// ── enigmagent_vault_status ───────────────────────────────────────────────────

export function createVaultStatusTool(client: VaultClient): AgentTool<Record<string, never>, VaultStatusToolResult> {
  return {
    name: 'enigmagent_vault_status',
    description:
      'Check whether the EnigmAgent vault server is running and the vault is unlocked. ' +
      'Call this before any operation that requires secrets to ensure the vault is ready. ' +
      'Returns running=true/false and unlocked=true/false.',
    inputSchema: {
      type: 'object',
      properties: {},
    },
    async execute(_input) {
      try {
        const status = await client.getStatus();
        const unlocked = status.unlocked;
        return {
          running: true,
          unlocked,
          message: unlocked
            ? 'Vault is running and unlocked. Secret references ({{NAME}}) will be resolved.'
            : 'Vault server is running but locked. Unlock it with: enigmagent-mcp --mode rest --vault <path>',
        };
      } catch (err) {
        if (err instanceof VaultError && err.code === 'server_unreachable') {
          return {
            running: false,
            unlocked: false,
            message:
              'EnigmAgent server is not running. Start it with: ' +
              'enigmagent-mcp --mode rest --port 3737 --vault ~/.enigmagent/vault.json',
          };
        }
        throw err;
      }
    },
  };
}

// ── enigmagent_vault_list ─────────────────────────────────────────────────────

export function createVaultListTool(client: VaultClient): AgentTool<Record<string, never>, VaultListToolResult> {
  return {
    name: 'enigmagent_vault_list',
    description:
      'List all secrets available in the EnigmAgent vault by name and domain binding. ' +
      'Never returns actual secret values. Use the names shown here as {{PLACEHOLDER}} ' +
      'references in tool parameters — they will be resolved automatically at execution time. ' +
      'Example: if a secret named GITHUB_TOKEN is listed, reference it as {{GITHUB_TOKEN}} ' +
      'in any bash command, HTTP header, or form field.',
    inputSchema: {
      type: 'object',
      properties: {},
    },
    async execute(_input) {
      const entries = await client.listSecrets();
      return {
        count: entries.length,
        entries: entries.map(({ name, domain, created }) => ({
          name,
          domain,
          created,
        })),
      };
    },
  };
}

// ── Convenience factory ───────────────────────────────────────────────────────

/**
 * Create all EnigmAgent tools pre-wired to a VaultClient.
 * Pass the result directly to `definePlugin({ tools })`.
 */
export function createAllTools(client: VaultClient): AgentTool[] {
  return [
    createVaultStatusTool(client),
    createVaultListTool(client),
  ];
}
